### PR TITLE
Deprecate Tags.setState, and throw an exception when it is called after getState.

### PR DESCRIPTION
**Pull request by sebright**
This commit is related to #608.  It deprecates Tags.setState and
TagsComponent.setState and makes NoopTagsComponent.setState and
TagsComponentImplBase.setState throw IllegalStateException when they are called
after getState.

_________________________________________________________________

This PR is similar to #789, but applied to tagging.

### CODE REVIEW

File: core_impl/src/main/java/io/opencensus/implcore/tags/CurrentTaggingState.java [hunk 1]
```diff
@@ -29,13 +31,25 @@
...
+
+  @GuardedBy("this")
+  private boolean isRead;
+
+  public synchronized TaggingState get() {
+    isRead = true;
+    return getInternal();
+  }
+
+  public synchronized TaggingState getInternal() {
```
Comment (songy23) [inline core_impl/src/main/java/io/opencensus/implcore/tags/CurrentTaggingState.java]: This can be package-private.

Comment (sebright) [reply core_impl/src/main/java/io/opencensus/implcore/tags/CurrentTaggingState.java]: This method is called from TagContextBinarySerializerImpl, which is in a different package.

File: api/src/main/java/io/opencensus/tags/TagsComponent.java [hunk 2]
```diff
@@ -37,16 +37,25 @@
    * <p>When no implementation is available, {@code getState} always returns {@link
    * TaggingState#DISABLED}.
    *
+   * <p>Once {@link #getState()} is called, subsequent calls to {@link #setState(TaggingState)} will
```
Comment (bogdandrutu) [inline api/src/main/java/io/opencensus/tags/TagsComponent.java]: You can say that user of this API can cache the value.

File: core_impl/src/main/java/io/opencensus/implcore/tags/CurrentTaggingState.java [hunk 3]
```diff
@@ -29,13 +31,25 @@
...
-  private volatile TaggingState currentState = TaggingState.ENABLED;
 
-  public TaggingState get() {
+  @GuardedBy("this")
+  private TaggingState currentState = TaggingState.ENABLED;
+
+  @GuardedBy("this")
+  private boolean isRead;
+
+  public synchronized TaggingState get() {
```
Comment (bogdandrutu) [inline core_impl/src/main/java/io/opencensus/implcore/tags/CurrentTaggingState.java]: I think it is fine for the moment, but I would like to create an issue to remove this global lock.

File: impl_core/src/main/java/io/opencensus/implcore/tags/TagsComponentImplBase.java [hunk 4]
```diff
@@ -50,6 +50,7 @@ public TaggingState getState() {
   }
 
   @Override
+  @Deprecated
```
Comment (bogdandrutu) [inline impl_core/src/main/java/io/opencensus/implcore/tags/TagsComponentImplBase.java]: Same comment from stats. This is not visible so no need to mark as deprecated.

**Comment by sebright** [comment]
> I don't believe in deprecated internal methods, but up to you, I don't mind either way.

I'll leave it as it is, since it also means that I don't need to suppress a deprecation warning.

**Comment by sebright** [comment]
Thanks for the reviews!

**Pull request merged**